docs/tutorials/datastream_api.md

```diff
@@ -0,0 +1,252 @@
+
+### Basic stream sources
+
+The example above constructs a `DataStream<Person>` using `env.fromElements(...)`. This is a
+convenient way to throw together a simple stream for use in a prototype or test. There is also a
+`fromCollection(Collection)` method on `StreamExecutionEnvironment`. So instead, you could do this:
+
+{% highlight java %}
+List<Person> people = new ArrayList<Person>();
+
+people.add(new Person("Fred", 35));
+people.add(new Person("Wilma", 35));
+people.add(new Person("Pebbles", 2));
+
+DataStream<Person> flintstones = env.fromCollection(people);
+{% endhighlight %}
+
+Another convenient way to get some data into a stream while prototyping is to use a socket
+
+{% highlight java %}
+DataStream<String> lines = env.socketTextStream("localhost", 9999)
+{% endhighlight %}
+
+or a file
+
+{% highlight java %}
+DataStream<String> lines = env.readTextFile("file:///path");
+{% endhighlight %}
+
+In real applications the most commonly used data sources are those that support low-latency, high
+throughput parallel reads in combination with rewind and replay -- the prerequisites for high
+performance and fault tolerance -- such as Apache Kafka, Kinesis, and various filesystems. REST APIs
+and databases are also frequently used for stream enrichment.
+
+### Basic stream sinks
+
+The example above uses `adults.print()` to print its results to the task manager logs (which will
+appear in your IDE's console, when running in an IDE). This will call `toString()` on each element
+of the stream.
+
+The output looks something like this
+
+    1> Fred: age 35
+    2> Wilma: age 35
+
+where 1> and 2> indicate which sub-task (i.e., thread) produced the output.
+
+You can also write to a text file
+
+{% highlight java %}
+stream.writeAsText("/path/to/file")
+{% endhighlight %}
+
+or a CSV file
+
+{% highlight java %}
+stream.writeAsCsv("/path/to/file")
+{% endhighlight %}
+
+or a socket
+
+{% highlight java %}
+stream.writeToSocket(host, port, SerializationSchema)
+{% endhighlight %}
+
+In production, commonly used sinks include Kafka as well as various databases and filesystems.
+
+### Debugging
+
+In production your application will run in a remote cluster or set of containers. And if it fails,
+it will fail remotely. The Flink Master and Task Manager logs can be very helpful in debugging such
+failures, but it's much easier to do local debugging inside an IDE, which is something that Flink
+supports. You can set breakpoints, examine local variables, and step through your code. You can also
+step into Flink's code, which can be a great way to learn more about its internals if you are
+curious to see how Flink works.
+
+{% top %}
+
+## Hands-on
+
+At this point you know enough to get started coding and running a simple DataStream application.
+Clone the [flink-training repo](https://github.com/apache/flink-training), and after following the
+instructions in the README, do the first exercise. 
```

Comment:
Maybe rather than saying "first exercise", say which exercise ("Filtering a Stream (Ride Cleansing)") in case we re-order things. Or use a combination like "first exercise: Filtering a Stream (Ride Cleansing)"